pkg/http/server_group.go

```diff
@@ -0,0 +1,35 @@
+package http
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// NewServerGroup creates a new Server to start and gracefully stop a collection
+// of Servers.
+func NewServerGroup(servers ...Server) Server {
+	return &serverGroup{
+		servers: servers,
+	}
+}
+
+// serverGroup manages the starting and graceful shutdown of a collection of
+// servers.
+type serverGroup struct {
+	servers []Server
+}
+
+// Start runs the servers in the server group.
+func (s *serverGroup) Start(ctx context.Context) error {
+	g, groupCtx := errgroup.WithContext(ctx)
+
+	for _, server := range s.servers {
+		srv := server
+		g.Go(func() error {
+			return srv.Start(groupCtx)
+		})
+	}
+
+	return g.Wait()
```

Comment:
So what happens if only 1 set of servers in the errgroup fails? This waits for all Goroutines to finish right?

I would think if either main or prometheus (or there HTTP vs HTTPS variant if both are stood up) fail, we want the whole server to fail so we aren't in a partially failed state? (then let whatever platform hosting restart mechanisms are in place to start the whole suite up again with everything?)

Reply:
As soon as the first non-nil error is returned, the `groupCtx` is cancelled which causes the shut down of all of the servers we started. Checking the godoc for `errgroup.WithContext` as this is what I think it is supposed to do at least:

```
WithContext returns a new Group and an associated Context derived from ctx.

The derived Context is canceled the first time a function passed to Go returns a non-nil error or the first time Wait returns, whichever occurs first.
```